x/market/handler/handler_test.go

```diff
@@ -0,0 +1,203 @@
+	ptypes "github.com/ovrclk/akash/x/provider/types"
+)
+
+type testSuite struct {
+	t       *testing.T
+	ms      sdk.CommitMultiStore
+	ctx     sdk.Context
+	mkeeper keeper.Keeper
+	dkeeper dkeeper.Keeper
+	pkeeper pkeeper.Keeper
+	bkeeper bank.Keeper
+
+	handler sdk.Handler
+}
+
+func setupTestContext(t *testing.T) *testSuite {
+	suite := &testSuite{
+		t: t,
+	}
+
+	testutil.AccAddress(t)
+	mKey := sdk.NewKVStoreKey(types.StoreKey)
+	dKey := sdk.NewKVStoreKey(dtypes.StoreKey)
+	pKey := sdk.NewKVStoreKey(ptypes.StoreKey)
+
+	db := dbm.NewMemDB()
+	suite.ms = store.NewCommitMultiStore(db)
+	suite.ms.MountStoreWithDB(mKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(dKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(pKey, sdk.StoreTypeIAVL, db)
+
+	err := suite.ms.LoadLatestVersion()
+	require.NoError(t, err)
+
+	suite.ctx = sdk.NewContext(suite.ms, abci.Header{}, true, log.NewNopLogger())
+
+	suite.mkeeper = keeper.NewKeeper(app.MakeCodec(), dKey)
+	suite.dkeeper = dkeeper.NewKeeper(app.MakeCodec(), mKey)
+	suite.pkeeper = pkeeper.NewKeeper(app.MakeCodec(), pKey)
+
+	suite.handler = handler.NewHandler(handler.Keepers{
+		Market:     suite.mkeeper,
+		Deployment: suite.dkeeper,
+		Provider:   suite.pkeeper,
+		Bank:       suite.bkeeper,
+	})
+
+	return suite
+}
+
+func TestCreateBidValid(t *testing.T) {
+	suite := setupTestContext(t)
+
+	order, gspec := suite.createOrder(nil)
+
+	msg := types.MsgCreateBid{
+		Order:    order.ID(),
+		Provider: suite.createProvider(gspec.Requirements).Owner,
+		Price:    sdk.Coin{},
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.NoError(t, err)
+}
+
+func TestCreateBidNonExistingOrder(t *testing.T) {
+	suite := setupTestContext(t)
+
+	msg := types.MsgCreateBid{
+		Order:    types.OrderID{},
+		Provider: nil,
+		Price:    sdk.Coin{},
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.EqualError(t, err, types.ErrInvalidOrder.Error())
+}
+
+func TestCreateBidOverprice(t *testing.T) {
+	suite := setupTestContext(t)
+
+	res := []dtypes.Resource{
+		{
+			Price: sdk.NewCoin("thepricedcoin", sdk.NewInt(1)),
+		},
+	}
+	order, gspec := suite.createOrder(res)
+
+	msg := types.MsgCreateBid{
+		Order:    order.ID(),
+		Provider: suite.createProvider(gspec.Requirements).Owner,
+		Price:    sdk.NewCoin("thepricedcoin", sdk.NewInt(math.MaxInt64)),
+	}
+
+	_, err := suite.handler(suite.ctx, msg)
+	require.EqualError(t, err, types.ErrBidOverOrder.Error())
+}
+
+func TestCreateBidInvalidOrder(t *testing.T) {
+	testCtx := setupTestContext(t)
+
+	msg := types.MsgCreateBid{
+		Order:    types.OrderID{},
+		Provider: nil,
+		Price:    sdk.Coin{},
+	}
+
+	_, err := testCtx.handler(testCtx.ctx, msg)
+	require.EqualError(t, err, types.ErrInvalidOrder.Error())
+}
+
+func TestCreateBidInvalidAttributes(t *testing.T) {
+	suite := setupTestContext(t)
+
+	order, _ := suite.createOrder(nil)
+
+	msg := types.MsgCreateBid{
+		Order:    order.ID(),
+		Provider: suite.createProvider(nil).Owner,
```

Comment:
can use `testutil.Attributes(t)` instead of `nil` for the attributes - might test more